imports/lib/publications/TypedPublication.ts

```diff
@@ -0,0 +1,142 @@
+  } & (
+    Args extends void ? { validate?: undefined } : { validate: TypedPublicationValidator<Args> }
+  )) {
+    if (this.definition) {
+      throw new Error(`TypedPublication ${this.name} has already been defined`);
+    }
+
+    this.definition = {
+      validate: (validate ?? voidValidator) as TypedPublicationValidator<Args>,
+      run,
+    };
+
+    /* eslint-disable-next-line @typescript-eslint/no-this-alias -- we need to
+       capture both the TypedPublication object (so we can find the
+       implementation) and the method invocation context (for things like
+       userId and added/changed/removed) */
+    const self = this;
+    /* eslint-disable-next-line meteor/audit-argument-checks -- because we're
+      wrapping the method call, the eslint check can't tell what's going on,
+      but the actual Meteor audit-argument-checks package will still fire if
+      we don't use check methods. */
+    Meteor.publish(this.name, function (arg0: Args) {
+      return self.execute(this, arg0);
+    });
+  }
+
+  async execute(context: Subscription, arg0: unknown) {
+    if (!this.definition) {
+      throw new Error(`TypedPublication ${this.name} has not been defined`);
+    }
+
+    const { validate, run } = this.definition;
+    try {
+      const validatedArg0 = validate.bind(context)(arg0);
+      return await run.bind(context)(validatedArg0 as any);
+    } catch (error) {
+      Ansible.error('Error in publication', {
+        name: this.name,
+        user: context.userId,
+        arguments: arg0,
+        error: error instanceof Error ? error.message : error,
+      });
+      if (error instanceof Error && Bugsnag.isStarted()) {
+        Bugsnag.notify(error, (event) => {
+          // eslint-disable-next-line no-param-reassign
+          event.context = this.name ?? 'default publication';
+        });
+      }
+      throw error;
+    }
+  }
+}
+
+export class DefaultTypedPublication extends BaseTypedPublication<void> {
+  constructor() {
+    super(null);
+  }
+}
+
+export default class TypedPublication<
+  Args extends TypedPublicationArgs,
+> extends BaseTypedPublication<Args> {
+  override name!: string;
+
+  /* eslint-disable-next-line no-useless-constructor -- name is required */
+  constructor(name: string) {
+    super(name);
+  }
+
+  subscribe<T>(...args: TypedMethodSubscribeArgs<T, Args>): Meteor.SubscriptionHandle {
+    return Meteor.subscribe(this.name, ...args);
+  }
+
+  subscribeAsync<T>(...args: TypedMethodSubscribeArgs<T, Args>) {
+    return new Promise<Meteor.SubscriptionHandle>((resolve, reject) => {
+      const handle = Meteor.subscribe(this.name, ...args, {
+        onStop: (reason?: Meteor.Error) => {
+          if (reason) {
+            reject(reason);
+          }
+        },
+        onReady: () => {
+          resolve(handle);
+        },
+      });
+    });
+  }
+
+  useSubscribe<T>(...args: TypedMethodSubscribeArgs<T, Args>) {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
```

Comment:
I'm curious what makes this rule fire and if it makes sense to try to abide by its constraints or not.  Is this just the rule failing to infer this as something that should only be called where a hook can be called?

Reply:
If I'm reading this right, it trips any time you call a function name that looks like a hook inside of any class method: https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L509-L519

Reply:
Ahhh.  What do you think about shaping this as a standalone hook function that you pass the publication into instead?  That might be more in line with how React wants you to use hooks for testability, even though we don't really exercise that much?

Something like (probably needs type parameters worked out):

```typescript
function useTypedSubscribe<T>(publication: TypedPublication<T>, ...args: TypedMethodSubscribeArgs<T, Args>) {
  useSubscribe(pub.name, ...args);
}
```

I'm also fine with this as it is, just exploring the design space a little.

Reply:
Yeah, that seems like a good pattern. I'll switch to that.

Reply:
I think that also gives us a slightly nicer path forward on conditional subscribes - we can just make the publication an optional parameter.

Reply:
Yeah that turned out well -- winds up feeling very similar to the existing `useSubscribe` at the callsite.